From 07438eb16c408eec0bf23d8c3adb8e568f5471f7 Mon Sep 17 00:00:00 2001
From: Kody A Kantor <kody@kkantor.com>
Date: Thu, 26 Jul 2018 20:29:50 +0000
Subject: [PATCH] MANTA-3808 add functionality to disable mackerel and mola
 jobs MANTA-3812 remove 'manta_pg_job' from mola MANTA-3823 mola's moray_gc
 needs to specify a bucket name

---
 bin/gc_create_links.js       |  40 +++++-
 bin/kick_off_audit.js        |   9 +-
 bin/kick_off_gc.js           |   9 +-
 bin/kick_off_mpu_cleanup.js  |  22 +++-
 bin/kick_off_mpu_gc.js       |  13 +-
 bin/kick_off_pg_job.js       | 242 -----------------------------------
 bin/kick_off_pg_transform.js |  12 +-
 bin/moray_gc.js              |  76 ++++++++++-
 deps/jsstyle                 |   2 +-
 docs/system-crons.md         |  41 ++++++
 lib/job_manager.js           |  57 +++++++--
 sapi_manifests/mola/template |  30 ++++-
 12 files changed, 275 insertions(+), 278 deletions(-)
 delete mode 100755 bin/kick_off_pg_job.js

diff --git a/bin/gc_create_links.js b/bin/gc_create_links.js
index 19f6120..cc79406 100755
--- a/bin/gc_create_links.js
+++ b/bin/gc_create_links.js
@@ -7,7 +7,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -19,6 +19,7 @@ require('../lib/maxsockets')(256);
 
 var assert = require('assert-plus');
 var bunyan = require('bunyan');
+var fs = require('fs');
 var common = require('../lib').common;
 var getopt = require('posix-getopt');
 var manta = require('manta');
@@ -26,8 +27,9 @@ var path = require('path');
 var vasync = require('vasync');
 var stream = require('stream');
 var lstream = require('lstream');
+var verror = require('verror');
 
-var VE = require('verror').VError;
+var VE = verror.VError;
 
 
 var NAME = 'moray_gc_create_links';
@@ -38,6 +40,9 @@ var LOG = bunyan.createLogger({
 });
 var MANTA_CONFIG = (process.env.MANTA_CONFIG ||
                     '/opt/smartdc/common/etc/config.json');
+var MOLA_CONFIG = (process.env.MOLA_CONFIG ||
+                    '/opt/smartdc/mola/etc/config.json');
+var MOLA_CONFIG_OBJ = JSON.parse(fs.readFileSync(MOLA_CONFIG));
 var MANTA_CLIENT = manta.createClientFromFileSync(MANTA_CONFIG, LOG);
 var MANTA_USER = MANTA_CLIENT.user;
 var MANTA_DIR = '/' + MANTA_USER + '/stor/manta_gc/all/do';
@@ -49,12 +54,13 @@ var AUDIT = {
         'startTime': new Date()
 };
 var DIR_CACHE = {};
+var JOB_DISABLED_ERR = 'JobDisabled';
 
 
 function parseOptions() {
         var option;
         var opts = {};
-        var parser = new getopt.BasicParser('d:', process.argv);
+        var parser = new getopt.BasicParser('d:F', process.argv);
 
         while ((option = parser.getopt()) !== undefined) {
                 if (option.error) {
@@ -65,6 +71,9 @@ function parseOptions() {
                 case 'd':
                         opts.mantaDir = option.optarg;
                         break;
+                case 'F':
+                        opts.forceRun = true;
+                        break;
                 default:
                         usage('Unknown option: ' + option.option);
                         break;
@@ -73,6 +82,8 @@ function parseOptions() {
 
         //Set up some defaults...
         opts.mantaDir = opts.mantaDir || MANTA_DIR;
+        opts.jobEnabled = MOLA_CONFIG_OBJ.gcEnabled;
+        opts.disableAllJobs = MOLA_CONFIG_OBJ.disableAllJobs;
 
         return (opts);
 }
@@ -84,6 +95,7 @@ function usage(msg) {
         }
         var str  = 'usage: ' + path.basename(process.argv[1]);
         str += ' [-d manta_directory]';
+        str += ' [-F force_run]';
         console.error(str);
         process.exit(1);
 }
@@ -301,6 +313,22 @@ function createGcLinks(opts, cb) {
                 'client': MANTA_CLIENT,
                 'dir': opts.mantaDir
         };
+
+        if (opts.forceRun) {
+                LOG.info('Forcing job run');
+        } else {
+                if (opts.disableAllJobs === true) {
+                        cb(new VE({ 'name': JOB_DISABLED_ERR },
+                                'all jobs are disabled'));
+                        return;
+                }
+                if (opts.jobEnabled === false) {
+                        cb(new VE({ 'name': JOB_DISABLED_ERR },
+                                'GC job is disabled'));
+                        return;
+                }
+        }
+
         common.getObjectsInDir(gopts, function (err, objs) {
                 if (err && err.code === 'ResourceNotFound') {
                         LOG.info('GC not ready yet: ' + opts.mantaDir +
@@ -336,7 +364,11 @@ var _opts = parseOptions();
 
 createGcLinks(_opts, function (err) {
         if (err) {
-                LOG.fatal(err, 'Error.');
+                if (verror.hasCauseWithName(err, JOB_DISABLED_ERR)) {
+                        LOG.info(err);
+                } else {
+                        LOG.fatal(err, 'Error.');
+                }
         } else {
                 AUDIT.cronFailed = 0;
         }
diff --git a/bin/kick_off_audit.js b/bin/kick_off_audit.js
index 7e6e39d..c63a961 100755
--- a/bin/kick_off_audit.js
+++ b/bin/kick_off_audit.js
@@ -6,7 +6,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
@@ -93,7 +93,7 @@ function parseOptions() {
         // command line, and use the defaults if all else fails.
         var opts = MOLA_AUDIT_CONFIG_OBJ;
         opts.shards = opts.shards || [];
-        var parser = new getopt.BasicParser('a:c:d:m:np:r:s:t', process.argv);
+        var parser = new getopt.BasicParser('a:c:d:m:np:r:s:tF', process.argv);
 
         while ((option = parser.getopt()) !== undefined) {
                 if (option.error) {
@@ -133,6 +133,9 @@ function parseOptions() {
                         opts.jobName = 'manta_audit_test';
                         opts.jobRoot = MP + '/manta_audit_test';
                         break;
+                case 'F':
+                        opts.forceRun = true;
+                        break;
                 default:
                         usage('Unknown option: ' + option.option);
                         break;
@@ -147,6 +150,7 @@ function parseOptions() {
         opts.assetFile = opts.assetFile ||
                 '/opt/smartdc/common/bundle/mola.tar.gz';
 
+        opts.jobEnabled = opts.auditEnabled;
         opts.auditMapDisk = opts.auditMapDisk || 32;
         opts.auditReduceMemory = opts.auditReduceMemory || 4096;
         opts.auditReduceDisk = opts.auditReduceDisk || 16;
@@ -168,6 +172,7 @@ function usage(msg) {
         str += ' [-r marlin_reducer_memory]';
         str += ' [-s manta_storage_id]';
         str += ' [-t output_to_test]';
+        str += ' [-F force_run]';
         console.error(str);
         process.exit(1);
 }
diff --git a/bin/kick_off_gc.js b/bin/kick_off_gc.js
index c483baa..6ba1aa4 100755
--- a/bin/kick_off_gc.js
+++ b/bin/kick_off_gc.js
@@ -6,7 +6,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var bunyan = require('bunyan');
@@ -109,7 +109,7 @@ function parseOptions() {
         // command line, and use the defaults if all else fails.
         var opts = MOLA_CONFIG_OBJ;
         opts.shards = opts.shards || [];
-        var parser = new getopt.BasicParser('a:d:g:m:no:p:r:t', process.argv);
+        var parser = new getopt.BasicParser('a:d:g:m:no:p:r:tF', process.argv);
 
         while ((option = parser.getopt()) !== undefined) {
                 if (option.error) {
@@ -149,6 +149,9 @@ function parseOptions() {
                         opts.jobName = 'manta_gc_test';
                         opts.jobRoot = MP + '/manta_gc_test';
                         break;
+                case 'F':
+                        opts.forceRun = true;
+                        break;
                 default:
                         usage('Unknown option: ' + option.option);
                         break;
@@ -164,6 +167,7 @@ function parseOptions() {
         opts.assetFile = opts.assetFile ||
                 '/opt/smartdc/common/bundle/mola.tar.gz';
 
+        opts.jobEnabled = opts.gcEnabled;
         opts.gcMapDisk = opts.gcMapDisk || 32;
         opts.gcReduceMemory = opts.gcReduceMemory || 8192;
         opts.gcReduceDisk = opts.gcReduceDisk || 32;
@@ -194,6 +198,7 @@ function usage(msg) {
         str += ' [-o object_id]';
         str += ' [-r marlin_reducer_memory]';
         str += ' [-t output_to_test]';
+        str += ' [-F force_run]';
         console.error(str);
         process.exit(1);
 }
diff --git a/bin/kick_off_mpu_cleanup.js b/bin/kick_off_mpu_cleanup.js
index bd06b98..a0e4d6a 100755
--- a/bin/kick_off_mpu_cleanup.js
+++ b/bin/kick_off_mpu_cleanup.js
@@ -23,7 +23,6 @@ var util = require('util');
 var vasync = require('vasync');
 var vstream = require('vstream');
 
-var lib = require('../lib');
 var mpu = require('../lib/mpu');
 
 /*
@@ -229,6 +228,12 @@ function getOptions() {
                               'will be used.',
                         helpArg: 'DIR'
                 },
+                {
+                        names: ['forceRun', 'F'],
+                        type: 'bool',
+                        help: 'Run this job regardless of whether or not ' +
+                              'this job is disabled.'
+                },
                 {
                         names: ['help', 'h'],
                         type: 'bool',
@@ -900,6 +905,21 @@ var toCleanup;
  */
 var completed;
 
+if (userOpts.forceRun === true) {
+        LOG.info('Forcing job run');
+} else {
+        if (CONFIG.disableAllJobs === true) {
+                LOG.info('All jobs are disabled, exiting.');
+                process.exit(exitCode);
+        }
+
+        if (CONFIG.gcEnabled === false) {
+                LOG.info('GC is disabled, exiting.');
+                process.exit(exitCode);
+        }
+}
+
+
 if (!userOpts.file) {
         /*
          * Typical production case. Possible fatal errors are:
diff --git a/bin/kick_off_mpu_gc.js b/bin/kick_off_mpu_gc.js
index bb46f97..98b9bca 100755
--- a/bin/kick_off_mpu_gc.js
+++ b/bin/kick_off_mpu_gc.js
@@ -159,9 +159,13 @@ function parseOptions() {
          */
         var opts = MOLA_CONFIG_OBJ;
         opts.shards = opts.shards || [];
-        var parser = new getopt.BasicParser('a:bd:g:s:m:no:p:r:t',
+        var parser = new getopt.BasicParser('a:bd:g:s:m:no:p:r:tF',
                                             process.argv);
-        while ((option = parser.getopt()) !== undefined && !option.error) {
+        while ((option = parser.getopt()) !== undefined) {
+                if (option.error) {
+                        usage();
+                }
+
                 switch (option.option) {
                 case 'a':
                         opts.assetFile = option.optarg;
@@ -197,6 +201,9 @@ function parseOptions() {
                         opts.jobName = 'manta_mpu_gc_test';
                         opts.jobRoot = MP + '/manta_mpu_gc_test';
                         break;
+                case 'F':
+                        opts.forceRun = true;
+                        break;
                 default:
                         usage('Unknown option: ' + option.option);
                         break;
@@ -211,6 +218,7 @@ function parseOptions() {
         opts.assetFile = opts.assetFile ||
                 '/opt/smartdc/common/bundle/mola.tar.gz';
 
+        opts.jobEnabled = opts.gcEnabled;
         opts.gcMapDisk = opts.gcMapDisk || 32;
         opts.gcReduceMemory = opts.gcReduceMemory || 8192;
         opts.gcReduceDisk = opts.gcReduceDisk || 32;
@@ -238,6 +246,7 @@ function usage(msg) {
         str += ' [-o object_id]';
         str += ' [-r marlin_reducer_memory]';
         str += ' [-t output_to_test]';
+        str += ' [-F force_run]';
         console.error(str);
         process.exit(1);
 }
diff --git a/bin/kick_off_pg_job.js b/bin/kick_off_pg_job.js
deleted file mode 100755
index b692107..0000000
--- a/bin/kick_off_pg_job.js
+++ /dev/null
@@ -1,242 +0,0 @@
-#!/usr/bin/env node
-/*
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
-
-/*
- * Copyright (c) 2017, Joyent, Inc.
- */
-
-var bunyan = require('bunyan');
-var fs = require('fs');
-var getopt = require('posix-getopt');
-var lib = require('../lib');
-var manta = require('manta');
-var path = require('path');
-
-
-
-///--- Global Objects
-
-var NAME = 'pg_job';
-var LOG = bunyan.createLogger({
-        level: (process.env.LOG_LEVEL || 'info'),
-        name: NAME,
-        stream: process.stdout
-});
-var MOLA_CONFIG = (process.env.MOLA_CONFIG ||
-                   '/opt/smartdc/mola/etc/config.json');
-var MOLA_CONFIG_OBJ = JSON.parse(fs.readFileSync(MOLA_CONFIG));
-var MANTA_CLIENT = manta.createClientFromFileSync(MOLA_CONFIG, LOG);
-var MANTA_USER = MANTA_CLIENT.user;
-
-
-
-///--- Global Constants
-
-var MP = '/' + MANTA_USER + '/stor';
-
-
-
-///--- Helpers
-
-/* BEGIN JSSTYLED */
-function getEnvCommon(opts) {
-        return (' \
-set -o pipefail && \
-cd /assets/ && gtar -xzf ' + opts.marlinPathToAsset + ' && cd mola && \
-');
-}
-/* END JSSTYLED */
-
-
-/* BEGIN JSSTYLED */
-function getMapCmd(opts) {
-        var cmd = getEnvCommon(opts) + ' \
-gzcat -f | ./build/node/bin/node ./bin/pg_transform.js | \
-   ' + opts.map + ' \
-';
-        return (cmd);
-}
-/* END JSSTYLED */
-
-
-/* BEGIN JSSTYLED */
-function getReduceCmd(opts, red) {
-        var cmd = getEnvCommon(opts) + ' \
-' + red + ' \
-';
-        return (cmd);
-}
-/* END JSSTYLED */
-
-
-function parseOptions() {
-        var option;
-        //First take what's in the config file, override what's on the
-        // command line, and use the defaults if all else fails.
-        var opts = MOLA_CONFIG_OBJ;
-        opts.shards = opts.shards || [];
-        opts.reduces = opts.reduces || [];
-        opts.tablePrefixes = opts.tablePrefixes || [];
-        var parser = new getopt.BasicParser('a:c:e:m:np:r:st:', process.argv);
-
-        while ((option = parser.getopt()) !== undefined) {
-                if (option.error) {
-                        usage();
-                }
-
-                switch (option.option) {
-                case 'a':
-                        opts.assetFile = option.optarg;
-                        break;
-                case 'c':
-                        opts.reduces.push(option.optarg);
-                        break;
-                case 'e':
-                        opts.numberReducers = lib.common.parseNumberOption(
-                            option.optarg, '-e', 1, null, usage);
-                        break;
-                case 'm':
-                        opts.shards.push(option.optarg);
-                        break;
-                case 'n':
-                        opts.noJobStart = true;
-                        break;
-                case 'p':
-                        opts.map = option.optarg;
-                        break;
-                case 'r':
-                        opts.pgJobReduceMemory = lib.common.parseNumberOption(
-                            option.optarg, '-r', 1, null, usage);
-                        break;
-                case 's':
-                        opts.readFromStdin = true;
-                        break;
-                case 't':
-                        opts.tablePrefixes.push(option.optarg);
-                        break;
-                default:
-                        usage('Unknown option: ' + option.option);
-                        break;
-                }
-        }
-
-        if (!opts.map && !opts.readFromStdin) {
-                usage('map or reading from stdin must be specified');
-        }
-
-        if (opts.tablePrefixes.length === 0) {
-                usage('table prefixes is required (perhaps manta-)?');
-        }
-
-        //Set up some defaults...
-        opts.jobName = opts.jobName || 'manta_pg_job';
-        opts.jobRoot = opts.jobRoot || MP + '/manta_pg_job';
-
-        opts.assetDir = opts.jobRoot + '/assets';
-        opts.assetObject = opts.assetDir + '/mola.tar.gz';
-        opts.assetFile = opts.assetFile ||
-                '/opt/smartdc/common/bundle/mola.tar.gz';
-
-        opts.pgJobReduceMemory = opts.pgJobReduceMemory || 4096;
-        opts.marlinPathToAsset = opts.assetObject.substring(1);
-        opts.marlinAssetObject = opts.assetObject;
-
-        opts.numberReducers = opts.numberReducers || 1;
-
-        return (opts);
-}
-
-
-function usage(msg) {
-        if (msg) {
-                console.error(msg);
-        }
-        var str  = 'usage: ' + path.basename(process.argv[1]);
-        str += ' [-a asset_file]';
-        str += ' [-c reduce_command (can be repeated)]';
-        str += ' [-e number of reducers]';
-        str += ' [-m moray_shard (can be repeated)]';
-        str += ' [-n no_job_start]';
-        str += ' [-p map_command (only once)]';
-        str += ' [-r marlin_reducer_memory]';
-        str += ' [-s read job from stdin]';
-        str += ' [-t table_prefix]';
-        console.error(str);
-        process.exit(1);
-}
-
-
-function getJob(opts, cb) {
-        if (opts.readFromStdin) {
-                var jobString = '';
-                process.stdin.setEncoding('utf8');
-                process.stdin.on('data', function (chunk) {
-                        jobString += chunk;
-                });
-
-                process.stdin.on('end', function () {
-                        try {
-                                var j = JSON.parse(jobString);
-                        } catch (e) {
-                                cb(e);
-                                return;
-                        }
-                        cb(null, j);
-                });
-                process.stdin.resume();
-        } else {
-                var job = {
-                        phases: []
-                };
-
-                job.phases.push({
-                        type: 'map',
-                        exec: getMapCmd(opts)
-                });
-                for (var i = 0; i < opts.reduces.length; ++i) {
-                        job.phases.push({
-                                type: 'reduce',
-                                count: opts.numberReducers,
-                                memory: opts.pgJobReduceMemory,
-                                exec: getReduceCmd(opts, opts.reduces[i])
-                        });
-                }
-                cb(null, job);
-        }
-}
-
-
-function getObjects(opts, cb) {
-        LOG.info({ opts: opts }, 'Running Test.');
-        var shards = opts.shards;
-
-        if (shards.length === 0) {
-                cb(new Error('No shards specified.'));
-                return;
-        }
-
-        lib.common.findObjectsForShards({
-                'log': LOG,
-                'shards': shards,
-                'client': MANTA_CLIENT,
-                'tablePrefixes': opts.tablePrefixes
-        }, cb);
-}
-
-
-///--- Main
-
-var _opts = parseOptions();
-
-_opts.getJobDefinition = getJob;
-_opts.getJobObjects = getObjects;
-
-var jobManager = lib.createJobManager(_opts, MANTA_CLIENT, LOG);
-jobManager.run(function () {
-        MANTA_CLIENT.close();
-        LOG.info('Done for now.');
-});
diff --git a/bin/kick_off_pg_transform.js b/bin/kick_off_pg_transform.js
index af30228..4ba947b 100755
--- a/bin/kick_off_pg_transform.js
+++ b/bin/kick_off_pg_transform.js
@@ -6,7 +6,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
@@ -81,7 +81,7 @@ function parseOptions() {
         // command line, and use the defaults if all else fails.
         var opts = MOLA_CONFIG_OBJ;
         opts.shards = opts.shards || [];
-        var parser = new getopt.BasicParser('a:b:m:no:p:y:', process.argv);
+        var parser = new getopt.BasicParser('a:b:m:no:p:y:F', process.argv);
 
         while ((option = parser.getopt()) !== undefined) {
                 if (option.error) {
@@ -112,6 +112,9 @@ function parseOptions() {
                         opts.pgMapMemory = lib.common.parseNumberOption(
                             option.optarg, '-y', 1, null, usage);
                         break;
+                case 'F':
+                        opts.forceRun = true;
+                        break;
                 default:
                         usage('Unknown option: ' + option.option);
                         break;
@@ -130,6 +133,7 @@ function parseOptions() {
         opts.assetFile = opts.assetFile ||
                 '/opt/smartdc/common/bundle/mola.tar.gz';
 
+        opts.jobEnabled = opts.pgEnabled;
         opts.marlinPathToAsset = opts.assetObject.substring(1);
         opts.marlinAssetObject = opts.assetObject;
 
@@ -166,6 +170,7 @@ function usage(msg) {
         str += ' [-m moray_shard (can be repeated)]';
         str += ' [-n no_job_start]';
         str += ' [-o output_directory_prefix (defaults to object location)]';
+        str += ' [-F force_run]';
         str += '';
         str += 'The backfill object overrides the shard list.';
         console.error(str);
@@ -213,6 +218,9 @@ function startJobForObject(opts, cb) {
         var jopts = {
                 'jobName': jobName,
                 'jobRoot': jjobRoot,
+                'jobEnabled': opts.jobEnabled,
+                'disableAllJobs': opts.disableAllJobs,
+                'forceRun': opts.forceRun,
                 'morayDumpObject': object,
                 'getJobObjects': function (_, subcb) {
                         return (subcb(null, [ object.path ]));
diff --git a/bin/moray_gc.js b/bin/moray_gc.js
index 98487f6..4331086 100755
--- a/bin/moray_gc.js
+++ b/bin/moray_gc.js
@@ -23,8 +23,10 @@ var assert = require('assert-plus');
 var bunyan = require('bunyan');
 var exec = require('child_process').exec;
 var fs = require('fs');
+var getopt = require('posix-getopt');
 var lib = require('../lib');
 var manta = require('manta');
+var path = require('path');
 var vasync = require('vasync');
 var verror = require('verror');
 var stream = require('stream');
@@ -40,17 +42,15 @@ var LOG = bunyan.createLogger({
 });
 var MANTA_CONFIG = (process.env.MANTA_CONFIG ||
                     '/opt/smartdc/common/etc/config.json');
+var MOLA_CONFIG = (process.env.MOLA_CONFIG ||
+                    '/opt/smartdc/mola/etc/config.json');
+var MOLA_CONFIG_OBJ = JSON.parse(fs.readFileSync(MOLA_CONFIG));
 var MANTA_CLIENT = manta.createClientFromFileSync(MANTA_CONFIG, LOG);
 var MANTA_USER = MANTA_CLIENT.user;
 var MORAY_CLEANUP_PATH = '/' + MANTA_USER + '/stor/manta_gc/moray';
 var PID_FILE = '/var/tmp/moray_gc.pid';
 var CRON_START = new Date();
-var MORAY_CLEANER = lib.createMorayCleaner({ log: LOG, batchSize: 1000 });
-MORAY_CLEANER.on('error', function (err) {
-        LOG.fatal(err);
-        var returnCode = auditCron(err);
-        process.exit(returnCode);
-});
+
 
 
 /*
@@ -80,7 +80,8 @@ function cleanShardOneObject(log, shard, input, cb) {
                  */
                 var mcs = MORAY_CLEANER.cleanStream({
                         shard: shard,
-                        object: input
+                        object: input,
+                        bucket: 'manta_delete_log'
                 });
 
                 mcs.once('error', function (mcsErr) {
@@ -356,6 +357,67 @@ function auditCron(err) {
         return (audit.cronFailed);
 }
 
+function parseOptions() {
+        var option;
+        var opts = {};
+        var parser = new getopt.BasicParser('F', process.argv);
+        while ((option = parser.getopt()) !== undefined) {
+                if (option.error) {
+                        usage();
+                }
+
+                switch (option.option) {
+                case 'F':
+                        opts.forceRun = true;
+                        break;
+                default:
+                        usage('Unknown option: ' + option.option);
+                        break;
+                }
+        }
+
+        opts.jobEnabled = MOLA_CONFIG_OBJ.gcEnabled;
+        opts.disableAllJobs = MOLA_CONFIG_OBJ.disableAllJobs;
+
+        return (opts);
+}
+
+function usage(msg) {
+        if (msg) {
+                console.error(msg);
+        }
+
+        var str  = 'usage: ' + path.basename(process.argv[1]);
+        str += ' [-F force_run]';
+
+        console.error(str);
+        process.exit(1);
+}
+
+///--- Main
+
+var _opts = parseOptions();
+
+if (_opts.forceRun) {
+        LOG.info('Forcing job run');
+} else {
+        if (_opts.disableAllJobs === true) {
+                LOG.info('All jobs are disabled, exiting.');
+                process.exit(0);
+        }
+
+        if (_opts.jobEnabled === false) {
+                LOG.info('GC is disabled, exiting.');
+                process.exit(0);
+        }
+}
+
+var MORAY_CLEANER = lib.createMorayCleaner({ log: LOG, batchSize: 1000 });
+MORAY_CLEANER.on('error', function (err) {
+        LOG.fatal(err);
+        var returnCode = auditCron(err);
+        process.exit(returnCode);
+});
 
 checkAlreadyRunning(function (err) {
         if (err) {
diff --git a/deps/jsstyle b/deps/jsstyle
index da42b50..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit da42b50ceb12d431437b32efd4c411a8e2fac0c8
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
diff --git a/docs/system-crons.md b/docs/system-crons.md
index 978652e..9aa8d5c 100644
--- a/docs/system-crons.md
+++ b/docs/system-crons.md
@@ -129,3 +129,44 @@ the [cron configuration][cron] and the [manifest file][manifest].
 | (daily-metering)        | meter-previous-day.sh    | metering    |      01:00 |
 
 Also see MANTA-2438.
+
+# Administration
+
+Mola and Mackerel cron jobs can be disabled on an individual or global basis.
+This can be done from the headnode using `sapiadm`. All jobs are enabled by
+default.
+
+All jobs (both mackerel and mola) can be disabled by setting `DISABLE_ALL_JOBS`
+to 'true' in the 'ops' SAPI service, like so:
+```
+$ sapiadm update $(sdc-sapi /services?name=ops | json -Ha uuid) metadata.DISABLE_ALL_JOBS=true
+```
+
+Jobs can be disabled on an individual basis as well. These are the fields that
+can be set to either 'true' or 'false' to disable or enable jobs:
+
+| Manta Job Name                              | Enable/Disable field   |
+| ------------------------------------------- | ---------------------- |
+| all                                         | DISABLE_ALL_JOBS       |
+| audit                                       | AUDIT_ENABLED          |
+| gc, gc-links, moray-gc, mpu-gc, mpu-cleanup | GC_ENABLED             |
+| sql-to-json                                 | PG_ENABLED             |
+| storage-hourly-metering                     | METER_STORAGE_ENABLED  |
+| compute-hourly-metering                     | METER_COMPUTE_ENABLED  |
+| request-hourly-metering                     | METER_REQUEST_ENABLED  |
+| (daily-metering)                            | METER_PREV_DAY_ENABLED |
+
+For example, to disable all of the GC-family jobs you could run this command
+from the headnode:
+```
+$ sapiadm update $(sdc-sapi /services?name=ops | json -Ha uuid) metadata.GC_ENABLED=false
+```
+
+And if you decide to later re-enable GC you can run the same command setting
+`GC_ENABLED` to 'true':
+```
+$ sapiadm update $(sdc-sapi /services?name=ops | json -Ha uuid) metadata.GC_ENABLED=true
+```
+
+The change will be written to the mola or mackerel config files the next time
+that the in-zone config-agent polls SAPI.
diff --git a/lib/job_manager.js b/lib/job_manager.js
index e1603d0..5b1c46c 100644
--- a/lib/job_manager.js
+++ b/lib/job_manager.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
@@ -17,6 +17,7 @@ var MemoryStream = require('memorystream');
 var path = require('path');
 var util = require('util');
 var vasync = require('vasync');
+var verror = require('verror');
 
 
 
@@ -113,6 +114,29 @@ JobManager.prototype.run = function run(cb) {
                 'startTime': new Date()
         };
 
+
+        var jobDisabled = 'JobDisabled';
+        function stopIfDisabled(m_opts, subcb) {
+                if (m_opts.forceRun) {
+                        self.log.info('Forcing job run');
+                        subcb();
+                        return;
+                }
+                if (m_opts.disableAllJobs === true) {
+                        subcb(new verror.VError({ 'name': jobDisabled },
+                                'all jobs are disabled'));
+                        return;
+                }
+
+                if (m_opts.jobEnabled === false) {
+                        subcb(new verror.VError({ 'name': jobDisabled },
+                                '%s is disabled', m_opts.jobName));
+                        return;
+                }
+
+                subcb();
+        }
+
         function invokeGetJobObjects(_, subcb) {
                 opts.getJobObjects(opts, function (err, objects) {
                         if (err) {
@@ -173,8 +197,21 @@ JobManager.prototype.run = function run(cb) {
                 });
         }
 
+        function writeAuditRecord() {
+                if (opts.noJobStart === undefined ||
+                    opts.noJobStart === false) {
+                        var a = self.audit;
+                        a.endTime = new Date();
+                        a.cronRunMillis = (a.endTime.getTime() -
+                                           a.startTime.getTime());
+                        a.opts = self.opts;
+                        self.log.info(a, 'audit');
+                }
+        }
+
         vasync.pipeline({
                 'funcs': [
+                        stopIfDisabled,
                         self.auditPreviousJobs.bind(self),
                         self.checkRunningJobs.bind(self),
                         self.setupDirectories.bind(self),
@@ -185,6 +222,12 @@ JobManager.prototype.run = function run(cb) {
                 ],
                 'arg': this.opts
         }, function (err) {
+                if (err && verror.hasCauseWithName(err, jobDisabled)) {
+                        self.log.info(err);
+                        writeAuditRecord();
+                        cb();
+                        return;
+                }
                 if (err && (err.shouldNotFatal === undefined)) {
                         self.log.fatal(err, 'Error.');
                 } else {
@@ -197,16 +240,8 @@ JobManager.prototype.run = function run(cb) {
                                 self.log.info(err2, 'Error saving audit.');
                         }
 
-                        //Write out audit record.
-                        if (opts.noJobStart === undefined ||
-                            opts.noJobStart === false) {
-                                var a = self.audit;
-                                a.endTime = new Date();
-                                a.cronRunMillis = (a.endTime.getTime() -
-                                                   a.startTime.getTime());
-                                a.opts = self.opts;
-                                self.log.info(a, 'audit');
-                        }
+                        writeAuditRecord();
+
                         //Send back original error
                         cb(err);
                 });
diff --git a/sapi_manifests/mola/template b/sapi_manifests/mola/template
index 81e1393..33e31e4 100644
--- a/sapi_manifests/mola/template
+++ b/sapi_manifests/mola/template
@@ -31,16 +31,38 @@
     "auditMapDisk": {{AUDIT_MAP_DISK}}{{/AUDIT_MAP_DISK}}{{#AUDIT_REDUCE_DISK}},
     "auditReduceDisk": {{AUDIT_REDUCE_DISK}}{{/AUDIT_REDUCE_DISK}}{{#AUDIT_REDUCE_MEMORY}},
     "auditReduceMemory": {{AUDIT_REDUCE_MEMORY}}{{/AUDIT_REDUCE_MEMORY}}{{#AUDIT_REDUCER_COUNT}},
-    "auditReducerCount": {{AUDIT_REDUCER_COUNT}}{{/AUDIT_REDUCER_COUNT}}{{#CRUFT_MAP_DISK}},
+    "auditReducerCount": {{AUDIT_REDUCER_COUNT}}{{/AUDIT_REDUCER_COUNT}},
+    {{#AUDIT_ENABLED}}
+    "auditEnabled": {{AUDIT_ENABLED}},
+    {{/AUDIT_ENABLED}}
+    {{^AUDIT_ENABLED}}
+    "auditEnabled": true,
+    {{/AUDIT_ENABLED}}{{#CRUFT_MAP_DISK}}
     "cruftMapDisk": {{CRUFT_MAP_DISK}}{{/CRUFT_MAP_DISK}}{{#CRUFT_REDUCE_DISK}},
     "cruftReduceDisk": {{CRUFT_REDUCE_DISK}}{{/CRUFT_REDUCE_DISK}}{{#CRUFT_REDUCE_MEMORY}},
     "cruftReduceMemory": {{CRUFT_REDUCE_MEMORY}}{{/CRUFT_REDUCE_MEMORY}}{{#CRUFT_REDUCER_COUNT}},
     "cruftReducerCount": {{CRUFT_REDUCER_COUNT}}{{/CRUFT_REDUCER_COUNT}}{{#GC_MAP_DISK}},
     "gcMapDisk": {{GC_MAP_DISK}}{{/GC_MAP_DISK}}{{#GC_REDUCE_DISK}},
     "gcReduceDisk": {{GC_REDUCE_DISK}}{{/GC_REDUCE_DISK}}{{#GC_REDUCE_MEMORY}},
-    "gcReduceMemory": {{GC_REDUCE_MEMORY}}{{/GC_REDUCE_MEMORY}}{{#PG_JOB_REDUCE_MEMORY}},
-    "pgJobReduceMemory": {{PG_JOB_REDUCE_MEMORY}}{{/PG_JOB_REDUCE_MEMORY}}{{#REBALANCE_MEMORY}},
+    "gcReduceMemory": {{GC_REDUCE_MEMORY}}{{/GC_REDUCE_MEMORY}},
+    {{#GC_ENABLED}}
+    "gcEnabled": {{GC_ENABLED}},
+    {{/GC_ENABLED}}
+    {{^GC_ENABLED}}
+    "gcEnabled": true,
+    {{/GC_ENABLED}}{{#REBALANCE_MEMORY}}
     "rebalanceMemory": {{REBALANCE_MEMORY}}{{/REBALANCE_MEMORY}}{{#PG_MAP_DISK}},
     "pgMapDisk": {{PG_MAP_DISK}}{{/PG_MAP_DISK}}{{#PG_MAP_MEMORY}},
-    "pgMapMemory": {{PG_MAP_MEMORY}}{{/PG_MAP_MEMORY}}
+    "pgMapMemory": {{PG_MAP_MEMORY}}{{/PG_MAP_MEMORY}},
+    {{#PG_ENABLED}}
+    "pgEnabled": {{PG_ENABLED}},
+    {{/PG_ENABLED}}
+    {{^PG_ENABLED}}
+    "pgEnabled": true,
+    {{/PG_ENABLED}}
+    {{#DISABLE_ALL_JOBS}}
+    "disableAllJobs": {{DISABLE_ALL_JOBS}}
+    {{/DISABLE_ALL_JOBS}}{{^DISABLE_ALL_JOBS}}
+    "disableAllJobs": false
+    {{/DISABLE_ALL_JOBS}}
 }
-- 
2.21.0

